paddle/framework/net_design.md

```diff
@@ -0,0 +1,159 @@
+  std::vector<string> outputs_;
+};
+```
+
+All network implementations should build networks from  a protobuf message which 
+describes the structure of a real network; `Run` method should be implemented by 
+all implementations to offer a universal method to forward or backward compute a network.
+
+A method of factory pattern can be defined like
+
+```c++
+std::unique<NetworkBase* CreateNet(const NetDef& def) {
+  switch (def.model_type()) {
+    case NN:
+      return new Network(def);
+    case Recursive:
+      return new RecursiveNet(def);
+    case Recurrent:
+      return new RecurrentNet(def);
+  }
+  return new Network(def);
+}
+```
+
+Network is designed as the container of operators, to make it more extendable,
+we decompling it from the related variable resources. 
+A `scope` is provided to `Run` so that the network structure can be reused 
+in different scopes.
+
+Finally, `NetworkBase` can be used as followed
+
+```c++
+Scope default_scope;
+auto net = CreateNet(def);
+net.Run(&default_scope);
+```
+
+
+## A Simple Network Implemention
+
+A very basic implemention is as followed, all it does is simply to run every operators in sequence.
+
+```c++
+class ScratchNet final : public NetworkBase {
+ public:
+  // Create a network describe by `def`.  NetDef is the definition of a network.
+  ScratchNet(const NetDef &def);
+
+  // Run all the operators with the `scope`, if no scope is provided, default
+  // scope will be used instead.
+  virtual bool Run(Scope *scope = nullptr) const override;
+
+ protected:
+  // Create operators accordding to `def`.
+  bool CreateNet(const NetDef &def);
+
+  // Add a operator which is identified as `type` and has attributes described
+  // in `attrs`, the `inputs` are the keys of readonly input variables,
+  // `outputs` are keys of mutable output variables.
+  bool AddOp(const std::string &type, const std::vector<string> &inputs,
+             const std::vector<string> &outputs,
+             const OprAttr &attrs = OprAttr());
+
+ private:
+  // the operators owned by `Network`.
+  std::vector<std::unique_ptr<Operator>> ops_;
+};
+```
+
+`ScratchNet` will create operators so that a private member `ops_` is defined,
+the operators are created by `CreateNet`, and each operator is created by `AddOp`.
+
+
+## Usage
+`ScratchNet` can be used to define and run a network as followed
+
+```c++
+// create an empty scope located on CPU device.
+Scope scope(CPUPlace());
+
+// create and init variables described in `net_desc`.
+scope.CreateVariables(net_desc);
+scope.InitVariables(net_desc);
+
+// create a network according to `net_desc`
+auto net = CreateNet(net_desc);
+
+// run the network providing the `scope`.
+net.Run(&scope);
+```
+
+## Compatibility with RNN
+
+Benefit from the decoupling of `ScratchNet.Run` and `Scope`, `ScratchNet` is compatible with future RNN design, 
+for example we can implement a simple recurrent neural network as followed
+
+```c++
```

Comment:
Well, this part should be rewritten.

`Scope` contains a `parent` argument. The local scope can access the parent scope variables. It should not copy variables between scopes.

```cpp
Scope parent;

// in RNN
NetworkBase& stepNet;
std::vector<Scope> timestepScopes;
for (size_t i=0; i < seqLen; ++i) {
  timestepScopes.push_back(Scope(&parent));
  stepNet.Run(&timestepScopes);
}
```